contracts/SpokePool.sol

```diff
@@ -1,16 +1,176 @@
-//SPDX-License-Identifier: Unlicense
+// SPDX-License-Identifier: GPL-3.0-only
 pragma solidity ^0.8.0;
 
-contract SpokePool {
-    //   constructor() {}
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@uma/core/contracts/common/implementation/Testable.sol";
+import "@uma/core/contracts/common/implementation/Lockable.sol";
+import "@uma/core/contracts/common/implementation/MultiCaller.sol";
 
+interface WETH9Like {
+    function deposit() external payable;
+}
+
+/**
+ * @title SpokePool
+ * @notice Contract deployed on source and destination chains enabling depositors to transfer assets from source to 
+ * destination. Deposit orders are fulfilled by off-chain relayers who also interact with this contract. Deposited
+ * tokens are locked on the source chain and relayers send the recipient the desired token currency and amount
+ * on the destination chain. Locked source chain tokens are later sent over the canonical token bridge to L1.
+ */
+abstract contract SpokePool is Testable, Lockable, MultiCaller {
+    using SafeERC20 for IERC20;
+
+    // Timestamp when contract was constructed. Relays cannot have a quote time before this.
+    uint64 public deploymentTime;
+
+    // Track the total number of deposits. Used as a unique identifier for deposits.
+    uint256 public numberOfDeposits;
+
+    struct DestinationToken {
+        address token;
+        bool isWethToken;
+        bool depositsEnabled;
+    }
+
+    // Whitelist of origin token to destination token routings.
+    mapping(address => mapping(uint256 => DestinationToken)) public whitelistedDestinationRoutes;
+
+
+    /****************************************
+     *                EVENTS                *
+     ****************************************/
+    event WhitelistRoute(address originToken, uint256 destinationChainId, address destinationToken, bool isWethToken);
+    event DepositsEnabled(address originToken, uint256 destinationChainId, bool depositsEnabled);
+    event FundsDeposited(
+        uint256 nonce,
+        uint256 destinationChainId,
+        uint256 originChainId,
+        uint256 amount,
+        uint64 relayerFeePct,
+        uint64 quoteTimestamp,
+        address originToken,
+        address recipient,
+        address sender,
+        address destinationToken
```

Comment:
ye, this is a good point. we dont actually need to know the destination token address. really, we just need to know that the `originToken` is whitelisted to send to a particular destination chainID. 

At the most extreme we simply have `whitelistedDestinationRoutes` being :
```
// mapping of origin token to destination chainID to bool signalling if it enabled (or not).
mapping(address => mapping(uint256 => bool))
```

we then store the weth address that the spokePool is deployed on in a separate variable and using this we check for any weth related logic.

thoughts?

Reply:
We'd need to map chain Id to weth address right? I think this is possible not to include destinationToken. 

For a given deposit with origin token and destination chain ID, how would the relayer know which currency to relay in?

Reply:
After talking offline i've decided to change the mapping to `originTokenAddress => destChainId => destChainToken`. 
Also, `wethAddress` is stored in contract state

Reply:
So I still think this is a bit confusing. Is there no central source of truth for mainnet token address 0x1234 <=> optimism token address 0x5678?

If you consider the deposits _alone_, you only need to know the mapping between a token on the source and destination. However, the data worker, DVM voters, OO disputers, etc must have some central understanding of what mainnet token in the HubPool maps to which tokens on each destination chain. Otherwise, there's no way to do the rebalancing actions. If we have each destination chain defining these source destination mappings locally, we could technically end up with conflicting sources of truth with no concrete way to resolve.

I also don't quite see why destination chains need this info at all? It seems like this mapping could be stored on L1 rather than duplicated on each of these chains, but I could be missing the use case.

Reply:
yeah bringing the data worker requirements into context makes this more apparent. I agree we should store this exact mapping (`originTokenAddress => destChainId => destChainToken`) on L1 and then change the last part of the mapping to a `bool` on L2.